### ISSUE TITLE
Extended "Custom Controller" for GearVR/Daydream Controllers

### ISSUE THREAD

**Issue by stlewis** [c-frame/aframe-extras#207]
I've been playing around with the `universal-controls` component for a bit, and determined that none of the presets work well for a fairly basic use-case: Using the trackpad on a 3dof controller for movement.  The reason for this, as far as I can tell is that the 3dof controllers respond to a set of events that just aren't implemented anywhere else. To solve that, I suggest the following implementation that I've tested explicitly with a GearVR controller, but that should work with Daydream as well:

```javascript
AFRAME.registerComponent('universal-gearvr-controls', {
  schema: {
    enabled: { default: true }
  },

  init: function () {
    this.dVelocity = new THREE.Vector3();
    this.bindMethods();
  },

  play: function () {
    AFRAME.log('Adding Listeners')
    this.addEventListeners();
  },

  pause: function () {
    this.removeEventListeners();
    this.dVelocity.set(0, 0, 0);
  },

  remove: function () {
    this.pause();
  },

  addEventListeners: function () {
    const sceneEl = this.el.sceneEl;

    sceneEl.addEventListener('trackpadtouchstart', this.onTouchStart);
    sceneEl.addEventListener('trackpadtouchend', this.onTouchEnd);
  },

  removeEventListeners: function () {
    const sceneEl = this.el.sceneEl;

    sceneEl.removeEventListener('trackpadtouchstart', this.onTouchStart);
    sceneEl.removeEventListener('trackpadtouchend', this.onTouchEnd);
  },

  isVelocityActive: function () {
    return this.data.enabled && this.isMoving;
  },

  getVelocityDelta: function () {
    this.dVelocity.z = this.isMoving ? -1 : 0;
    return this.dVelocity.clone();
  },

  bindMethods: function () {
    this.onTouchStart = this.onTouchStart.bind(this);
    this.onTouchEnd = this.onTouchEnd.bind(this);
  },

  onTouchStart: function (e) {
    this.isMoving = true;
    e.preventDefault();
  },

  onTouchEnd: function (e) {
    this.isMoving = false;
    e.preventDefault();
  }
});

```

Up to date version can be found here: [https://github.com/stlewis/aframe-experiments/blob/master/components/universal-gearvr-controls.js](https://github.com/stlewis/aframe-experiments/blob/master/components/universal-gearvr-controls.js)

If I can confirm at some point that this _does_ work with Daydream, I'll probably rename the component `3dof-controls`.

**Comment by donmccurdy** [comment]
Sorry for the delay on this. I've just finished a big overhaul of `universal-controls`, and specifically it is renamed to `movement-controls`. If the component above is working for you, would you be interested in doing a PR? I think `trackpad-controls` might be a good name.

**Comment by donmccurdy** [comment]
For overview of changes, see https://github.com/donmccurdy/aframe-extras/tree/master/src/controls

**Comment by stlewis** [comment]
I actually just upgraded my Gear VR, so I’ve been eager to try my A-Frame stuff out in it! 😃

 

Let me take a minute to make sure it’s all working properly, then I’ll rename it and submit a PR.

 

Thanks!

**Comment by donmccurdy** [comment]
Fixed by #218.
